### ISSUE TITLE
Cannot log in to Lagoon logs UI

### ISSUE THREAD

**Issue by smlx** [uselagoon/lagoon#3161]
# Describe the bug

The Lagoon logs UI (Kibana or Opensearch Dashboards) gets a cookie from Keycloak which passes role information to the logs UI in order to enforce index viewing permissions.

If the user is in too many groups, this cookie can get so large that it reaches the [de facto browser limit of 4096 bytes](https://stackoverflow.com/q/52203972) and is rejected by the browser.

This causes a redirect loop when trying to log in to the logs UI.

## To Reproduce

1. Add a user to lots of projects in Lagoon
2. User visits the logs UI
3. User cannot login and encounters a "too many redirects" error.

## Expected behavior

Lagoon should avoid creating such a large cookie.

## Screenshots

n/a

## Additional context

In order to add users to a project a user must be `OWNER` within a project group. This means that a Lagoon user who is `OWNER` in their main customer group which contains all their projects will often also want to be `OWNER` in individual project groups which will increase the likelihood of hitting this bug.

The code creating the cookie looks roughly like this:

```javascript
var ArrayList = Java.type("java.util.ArrayList");
var groupsAndRoles = new ArrayList();
var forEach = Array.prototype.forEach;
// add all groups the user is part of
forEach.call(user.getGroups().toArray(), function(group) {
	// remove the group role suffixes
	// lets check if the group has a parent if this is a child
	if(group.getFirstAttribute("type") == "role-subgroup") {
		var parent = group.getParent();
		if(parent.getFirstAttribute("type") == "project-default-group") {
			var projectIds = parent.getFirstAttribute("lagoon-projects");
			if(projectIds !== null) {
				forEach.call(projectIds.split(","), function(g) {
					groupsAndRoles.add("p" + g);
				});
				return;
			}
		}
	}
	var groupName = group.getName().replace(/-owner|-maintainer|-developer|-reporter|-guest/gi,"");
	groupsAndRoles.add(groupName);
	return;
});
// add all roles the user is part of
forEach.call(user.getRoleMappings().toArray(), function(role) {
	var roleName = role.getName();
	groupsAndRoles.add(roleName);
});
exports = groupsAndRoles;
```

This adds roles for all the projects that the user is an `OWNER` of even if the user is already an `OWNER` of a group that contains the project.

Could this code be improved to inspect the `lagoon-projects` annotation on the group and avoid adding the individual project roles where the user already has index view permissions for the project via a group? That would significantly reduce the cookie size in the common case outlined above.

**Comment by Schnitzel** [comment]
I think we should rather fix the issue that people need to be `OWNER` of a group and projects of that group at the same time, as this sounds to me like a bug.

**Comment by smlx** [comment]
Ok, but adding redundant project roles to the cookie when the user already has index view permissions on that project via a group also seems like a bug.

**Comment by tobybellwood** [comment]
So - we had a long chat about this one, and roughly came up with the following proposition (for the opensearch part):

The issue is that an individual user can get added to multiple project groups, and these groups are the ones that (primarily) are consuming the cookie storage length.

What if, instead, Lagoon created a personalised tenant in Opensearch for that user, and instead added the index permissions to that, so that keycloak, instead of passing all the project-groups, only returned the actual groups, and the personalised tenant. It's a small inversion of the logic.

Lagoon would have to be modified slightly to be able to create the personalised tenant when required (i.e. when someone adds a user directly to a project). Lagoon would still recognise this as a user>project addition, but an additional opensearch-specific process would run in the relevant resolver (and also on removal from the project)

The upside for these users is that they get somewhere they can store their kibana-specific stuff, for Lagoon it can continue to have it's project-specific permissions, for keycloak is that it no longer needs to pass across all the project groups, and for opensearch is that it only needs to handle proper groups, and the personalised tenants are only created for users that need them (i.e. on first assignment).


For the Lagoon part, yes, a group owner should be able to add users to a project within that group without having to add themselves first.

**Comment by Schnitzel** [comment]
> What if, instead, Lagoon created a personalised tenant in Opensearch for that user

Mhh are you sure you mean "tenant" and not "role"?
Because each tenant creates a new index in kibana which would make the indexes explode (this is the reason that we removed the tenants for the projects), plus also if each user has their own tenant multiple users cannot share their dashboard anymore if they are in the same group.

but yes, creating an individual role in kibana for each user with that users specific index permissions and tenant access would probably solve the problem overall.

**Comment by smlx** [comment]
I agree with @Schnitzel, more tenants (roles?) is confusing for users ("hey, where did my visualisation go!?") and adds extra load to Opensearch.

IMO if Lagoon avoids adding redundant project-specific roles to the cookie it will solve the problem in practice because that is the specific scenario where we currently see the problem. It's also a much smaller change, and we know it works because the current workaround for this issue is to remove the user from project groups until they can log in.

**Comment by tobybellwood** [comment]
This would have to be done in hand with removing the project-specific roles. The only personalised tenant+roles created would be for those people who don't use a group currently, so the numbers of roles shouldn't explode (in fact, they should reduce).

Not creating an alternative to the project-groups would render those users unable to view anything in kibana, hence this idea?

**Comment by shreddedbacon** [comment]
I have something that appears to work, it's in the test infrastructure now and playing around with groups/clearing cookies etc returns the expected results.

it creates 2 maps
* one map is projectid:custom-group (all the user created groups)
* one map is projectid:project-group (all the project default groups)

it removes all the `custom-group` projectids from the `project-group` map so that only unique ungrouped project roles remain

then goes over the remaining 2 maps to add all the unique roles for the user to the exported array

```javascript
var ArrayList = Java.type("java.util.ArrayList");
var HashMap = Java.type("java.util.HashMap");
var HashSet = Java.type("java.util.HashSet");

var groupsAndRoles = new ArrayList();
var groupProjectIds = new HashMap();
var projectGroupProjectIds = new HashMap();

var forEach = Array.prototype.forEach;
// add all groups the user is part of
forEach.call(user.getGroups().toArray(), function(group) {
	// remove the group role suffixes
	// lets check if the group has a parent if this is a child
	var groupName = group.getName().replace(/-owner|-maintainer|-developer|-reporter|-guest/gi,"");
	if(group.getFirstAttribute("type") == "role-subgroup") {
		var parent = group.getParent();
		var projectIds = parent.getFirstAttribute("lagoon-projects");
		if(parent.getFirstAttribute("type") == "project-default-group") {
			if(projectIds !== null) {
				forEach.call(projectIds.split(","), function(g) {
					projectGroupProjectIds.put(g, groupName)
				});
				return;
			}
		} else {
			if(projectIds !== null) {
				forEach.call(projectIds.split(","), function(g) {
					groupProjectIds.put(g, groupName)
				});
				return;
			}
		}
	}
	return;
});

// add all the unique project ids roles the user has, that aren't in an already existing group
projectGroupProjectIds.keySet().removeAll(groupProjectIds.keySet());
for each (var e in projectGroupProjectIds.keySet()) groupsAndRoles.add("p"+e);

// now add all the users groups
var uniqueGroups = new HashSet(groupProjectIds.values());
for each (var e in uniqueGroups) groupsAndRoles.add(e);

// add all roles the user is part of
forEach.call(user.getRoleMappings().toArray(), function(role) {
	var roleName = role.getName();
	groupsAndRoles.add(roleName);
});
exports = groupsAndRoles;
```

**Comment by tobybellwood** [comment]
fixed in #3319
